Scripts/Data/Collection/ObservableCounter.cs

```diff
@@ -0,0 +1,114 @@
+    /// </summary>
+    /// <typeparam name="TElement">The type of the elements to count.</typeparam>
+    /// <typeparam name="TEvent">The <see cref="UnityEvent"/> type to use.</typeparam>
+    public class ObservableCounter<TElement, TEvent> : MonoBehaviour where TEvent : UnityEvent<TElement>, new()
+    {
+        /// <summary>
+        /// Emitted when a searched element is found.
+        /// </summary>
+        public TEvent ElementAdded = new TEvent();
+        /// <summary>
+        /// Emitted when a searched element is not found.
+        /// </summary>
+        public TEvent ElementRemoved = new TEvent();
+
+        /// <summary>
+        /// The elements being counted.
+        /// </summary>
+        public Dictionary<TElement, int> ElementsCounter
+        {
+            get;
+            protected set;
+        } = new Dictionary<TElement, int>();
+
+        /// <summary>
+        /// Increases the count on the given element.
+        /// </summary>
+        /// <param name="element">The element to count.</param>
+        public virtual void IncreaseCount(TElement element)
+        {
+            if (!isActiveAndEnabled || element == null)
+            {
+                return;
+            }
+
+            if (!ElementsCounter.ContainsKey(element))
+            {
+                ElementsCounter.Add(element, 0);
+            }
+
+            ElementsCounter[element]++;
+            if (ElementsCounter[element] == 1)
+            {
+                ElementAdded?.Invoke(element);
+            }
+        }
+
+        /// <summary>
+        /// Decreases the count on the given element.
+        /// </summary>
+        /// <param name="element">The element to count.</param>
+        public virtual void DecreaseCount(TElement element)
+        {
+            int currentValue = 0;
+            if (!isActiveAndEnabled || element == null || !ElementsCounter.TryGetValue(element, out currentValue) || currentValue <= 0)
+            {
+                return;
+            }
+
+            ElementsCounter[element]--;
+            if (ElementsCounter[element] <= 0)
+            {
+                ElementsCounter.Remove(element);
+                ElementRemoved?.Invoke(element);
+            }
+        }
+
+        /// <summary>
+        /// Removes the element from the counter.
+        /// </summary>
+        /// <param name="element">The element to clear.</param>
+        public virtual void RemoveFromCount(TElement element)
+        {
+            if (!isActiveAndEnabled || element == null || !ElementsCounter.ContainsKey(element))
+            {
+                return;
+            }
+
+            ElementsCounter.Remove(element);
+            ElementRemoved?.Invoke(element);
+        }
+
+        /// <summary>
+        /// Clears all elements from the counter.
+        /// </summary>
+        public virtual void Clear()
+        {
+            foreach (KeyValuePair<TElement, int> items in ElementsCounter.ToList())
```

Comment:
```suggestion
            foreach (TElement element in ElementsCounter.Values.ToList())
```

Reply:
`ElementsCounter.Keys.ToList` as I want the Element, not the int :)

Reply:
👍